daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -328,20 +333,33 @@ final class ImmArray[+A] private (
 
   /** O(n) */
   override def equals(that: Any): Boolean = that match {
-    case thatArr: ImmArray[_] if length == thatArr.length =>
-      indices forall { i =>
-        uncheckedGet(i) == thatArr.uncheckedGet(i)
+    case thatArr: ImmArray[_] =>
+      var succ = length == thatArr.length
+      var i = start
+      var j = thatArr.start
+      val end = start + length
+      while (succ && i < end) {
+        succ = array(i) == thatArr.array(j)
```

Comment:
Is `==` the right thing here or do we need to call `equals`?

Reply:
In scala `x == y` is syntactic sugar for `x.equals(y)`.  
To get reference equality between `x` and `y` you have to use` x.eq(y)`.